rfc/composite_table_availability.pseudocode

```diff
@@ -0,0 +1,123 @@
+
+PhysicalTableSchema:
+	getFieldName(Column c)
+
+DataSource:
+	String getName()
+	Map<FieldName, Set<Segment>> raw segment data
+	Map<FieldName, List<Interval>> squashedFromSegments  // optional optimization
+
+	Set<FieldName> getFields() //union of all keys
+	List<Interval> getIntervals(FieldName fieldName)
+	Map<FieldName,List<Interval>> getIntervals(Set<FieldName>) // map slice from squashedMap
+	Set<SegmentId> getSegments(List<Interval>)  // Filter all segments by overlap on time
```

Comment:
The raw data from druid is in the form of a set of Segments where Segment contains an interval and a set of fields.  Segments should be atomic over their unit of time, at any given instant there is one 'winning' segment, and the schema it has is correct for that instant.  So Segments aren't a function on fields, only on time.  

I was imagining a crude way of doing this, going over every segment in values and filtering on time overlap.  An optimizing data structure could be used to cache this data.